ARROW-15582: [C++] Add support for registering standard Substrait functions

westonpace opened the pull request:
This picks up where #13285 has left off.  It mostly focuses on the Substrait->Arrow direction at the moment.  In addition, basic support is added for named tables.  This makes it possible to create unit tests that read from in-memory tables instead of requiring unit tests to do a scan.

The PR creates some utilities in `test_plan_builder.h` which allow for the construction of simple Substrait plans programmatically.  This is used to create unit tests for the function mapping.

The PR extracts id "ownership" out of the `ExtensionIdRegistry` and into its own `IdStorage` class.

The PR gets rid of `NestedExtensionIdRegistryImpl` and instead makes `ExtensionIdRegistryImpl` nested if `parent_ != nullptr`.

vibhatha commented on cpp/src/arrow/compute/exec/util.cc:
```diff
@@ -383,5 +383,25 @@ size_t ThreadIndexer::Check(size_t thread_index) {
+Status TableSinkNodeConsumer::Consume(ExecBatch batch) {
+  auto guard = consume_mutex_.Lock();
+  ARROW_ASSIGN_OR_RAISE(auto rb, batch.ToRecordBatch(schema_, pool_));
+  batches_.push_back(rb);
```
should we move this?

westonpace replied:
Yes.  Done.

vibhatha commented on cpp/src/arrow/engine/substrait/extension_set.cc:
```diff
@@ -129,18 +136,19 @@ Result<ExtensionSet> ExtensionSet::Make(
-                           " not found");
+    set.functions_[i] = {{function_ids[i].uri, function_ids[i].name}, "dummy"};
+
+    // if (auto rec = registry->GetFunction(function_ids[i])) {
+    //   set.functions_[i] = {rec->id, rec->function_name};
+    //   continue;
+    // }
+    // return Status::Invalid("Function ", function_ids[i].uri, "#", function_ids[i].name,
+    //                        " not found");
```
should we remove this or intended to use?

westonpace replied:
I've uncommented this and merged `FunctionMapping` into `ExtensionIdRegistry`.

vibhatha commented on cpp/src/arrow/engine/substrait/function_test.cc:
```diff
@@ -0,0 +1,156 @@
+       "-119",
+       int8()}};
+  CheckValidTestCases(valid_test_cases);
+  // const std::vector<FunctionTestCase> error_test_cases = {
+  //     {{kSubstraitArithmeticFunctionsUri, "add"},
+  //      {"ERROR", "127", "10"},
+  //      {nullptr, int8(), int8()}}};
```
should we remove this?

westonpace replied:
I uncommented this and implemented these cases.

westonpace commented:
I moved the `FunctionMapping` into `ExtensionIdRegistry` and reworked a few things.  I think this is ready for review now.  I did try to avoid unnecessary changes but there are a few cleanups in here as well.

@rtpsw this will likely have some (hopefully limited) impact on the UDF work you have ongoing.
@vibhatha I still need to think through what impacts this will have on aggregates.  I'll take a look at that on Monday.

rtpsw commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -703,6 +703,11 @@ TEST(Substrait, ExtensionSetFromPlan) {
+      },
+      {
+        "extension_uri_anchor": 18,
+        "uri": ")" + std::string(kSubstraitArithmeticFunctionsUri) +
```
I think it would be useful, at least for PyArrow's sake, to expose the string value of this constant (and similar ones) like `default_extension_types_uri()` does.

westonpace replied:
I converted these constants to std::string.

westonpace replied:
I'm not sure if that will make them available to python or not without the functions though.  I'll try that out later.

westonpace replied:
I considered adding these functions but, at the end of the day, this is information that the caller should be getting from Substrait and I don't know if Arrow needs to provide them.  I don't want to hold up this PR on this so we still think these functions are needed we can create a follow-up PR and have the discussion there.

westonpace replied:
Actually, I had to revert these back to `char *` to make the linter happy.  My thought  on a follow-up PR still holds I think.

westonpace commented on cpp/src/arrow/engine/substrait/serde.h:
```diff
@@ -111,7 +113,7 @@ ARROW_ENGINE_EXPORT Result<std::vector<compute::Declaration>> DeserializePlans(
 /// \return a vector of ExecNode declarations, one for each toplevel relation in the
 /// Substrait Plan
-ARROW_ENGINE_EXPORT Result<compute::ExecPlan> DeserializePlan(
+ARROW_ENGINE_EXPORT Result<std::shared_ptr<compute::ExecPlan>> DeserializePlan(
```
I moved the named table provider into the conversion options object being added by @jvanstraten 's PR here: #13537

Hopefully any new options can go into ConversionOptions and that will prevent future creep.

westonpace commented:
I rebased on top of #13537 so that I could move the named table provider into the conversion options.  I added support for aggregation function converters, at least for the Substrait -> Arrow direction.  I won't finish Arrow -> Substrait for aggregates due  ARROW-17239.

I still need to add tests for Substrait->Arrow aggregates and then I will try and get this PR merged.

westonpace commented:
I added tests for aggregate functions.  I only added sum/min/max/mean.  The count function will be a bit trickier ([ARROW-17061](https://issues.apache.org/jira/browse/ARROW-17061)).  I think this PR has grown enough however (perhaps a bit too much) and I'd like to just merge in what I have so far and address the rest once the Substrait functions have finalized a bit (or as needed to advance other work).

vibhatha commented on cpp/src/arrow/compute/exec/util.cc:
```diff
@@ -383,5 +383,25 @@ size_t ThreadIndexer::Check(size_t thread_index) {
   return thread_index;
 }
 
+Status TableSinkNodeConsumer::Init(const std::shared_ptr<Schema>& schema,
+                                   BackpressureControl* backpressure_control) {
+  // If the user is collecting into a table then backpressure is meaningless
```
Not related to this PR, but just a thought, should we move TableSourceNode to utils?

westonpace replied:
The `SinkNodeConsumer` objects can be useful to add to the public API because users specify them as part of options.  That's why I had to move this one since the previous location wasn't public and I couldn't use it in tests.

The `TableSourceNode` is a node and those stay in the private API so I don't think we need to move that one.

vibhatha commented on cpp/src/arrow/engine/substrait/expression_internal.cc:
```diff
@@ -41,6 +41,70 @@ namespace internal {
+                                const ConversionOptions& conversion_options) {
+  if (func.phase() != substrait::AggregationPhase::AGGREGATION_PHASE_INITIAL_TO_RESULT) {
+    return Status::NotImplemented(
+        "Unsupported aggregation phase.  Only INITIAL_TO_RESULT is supported");
```
nit: Not sure if this is possible, but can we inclue the function.phase() in the message?

westonpace replied:
That's a good idea.  I recall someone mentioning something about protobuf descriptors.  I'll see if I can figure out a good way to get this string value.

westonpace replied:
I added an `EnumToString` method and used it to print out the invalid invocation / phase that the user provided.  The error looks like this:

```
NotImplemented: Unsupported aggregation invocation 'AGGREGATION_INVOCATION_DISTINCT'.  Only AGGREGATION_INVOCATION_ALL is supported
```

If the user provides a completely invalid value I thought I would get "unknown" but it looks like the google protobuf converter changes it to unspecified so I get:

```
NotImplemented: Unsupported aggregation invocation 'AGGREGATION_INVOCATION_UNSPECIFIED'.  Only AGGREGATION_INVOCATION_ALL is supported
```

vibhatha replied:
Yes that’s better.

vibhatha commented on cpp/src/arrow/engine/substrait/expression_internal.cc:
```diff
@@ -41,6 +41,70 @@ namespace internal {
+  if (func.invocation() !=
+      substrait::AggregateFunction::AggregationInvocation::
+          AggregateFunction_AggregationInvocation_AGGREGATION_INVOCATION_ALL) {
+    return Status::NotImplemented(
+        "Unsupported aggregation invocation.  Only AGGREGATION_INVOCATION_ALL is "
+        "supported");
```
nit: same here. Just to inform with the tried phase and say it is not supported rather than saying only supporting a particular option only.

westonpace replied:
Added `EnumToString` here too

vibhatha commented:
@westonpace General comment on usage of `auto` with `ARROW_ASSIGN_OR_RAISE`. There is a mix of auto and variable type through out the PR. Is there any specific reason? I think there was a recent discussion on minimizing usage of `auto` in certain places where it is hard to deterimine the data type for the readability sake. Is it the case here?

vibhatha commented on cpp/src/arrow/engine/substrait/ext_test.cc:
```diff
@@ -158,10 +152,10 @@ TEST_P(ExtensionIdRegistryTest, ReregisterFunctions) {
-    auto id = Id{kArrowExtTypesUri, name};
-    ASSERT_RAISES(Invalid, registry->CanRegisterFunction(id, name.to_string()));
-    ASSERT_RAISES(Invalid, registry->RegisterFunction(id, name.to_string()));
+  for (Id function_id : kFunctionIds) {
```
Should this be `const Id& function_id` instead?

westonpace replied:
TBH, I don't really know.  Small things we pass by value (e.g. `int`).  A `string_view` is a pair of pointers and generally considered small enough that it is [better to pass by value](https://quuxplusone.github.io/blog/2021/11/09/pass-string-view-by-value/).

An `Id` is a pair of `string_view` objects.  Does that mean it is small enough?  I'm not sure what the exact rules are here but the existing usage seems to be treating it as a small value object and so I have tried to continue to do so.

Perhaps @bkietz for a second opinion / further guidelines.

bkietz replied:
For anything trivial(ly copyable, destructible, ...) by-reference and by-value should be interchangeable. Using it by value here is more readable so do that. The compiler will have no difficulty removing the trivial copy

vibhatha commented on cpp/src/arrow/engine/substrait/ext_test.cc:
```diff
@@ -198,11 +207,11 @@ TEST(ExtensionIdRegistryTest, RegisterTempFunctions) {
 
     for (util::string_view name : kTempFunctionNames) {
```
I know it is not part of your edit, should this be `const util::string_view& name` instead?

westonpace replied:
I linked an article earlier that talks about why it is better to pass `util::string_view` by value.  I find it is helpful to think of `string_view` as just a specialized `char*`.  So I wouldn't do `const char*& name` and I don't do `const util::string_view& name`

bkietz replied:
:+1:

vibhatha commented on cpp/src/arrow/engine/substrait/ext_test.cc:
```diff
@@ -248,22 +257,22 @@ TEST(ExtensionIdRegistryTest, RegisterNestedFunctions) {
-    ASSERT_OK(registry1->CanRegisterFunction(id1, name1.to_string()));
-    ASSERT_OK(registry1->RegisterFunction(id1, name1.to_string()));
+    ASSERT_OK(registry1->CanAddSubstraitCallToArrow(id1));
+    ASSERT_OK(registry1->AddSubstraitCallToArrow(id1, name1.to_string()));
```
nit: `name1` is this okay? since there is `name2` could it be renamed?

westonpace replied:
I don't understand.  Are you suggesting `name`/`name2` instead of `name1`/`name2`?

westonpace replied:
Ping

vibhatha replied:
Ah no I mean instead it could be an array `name[0]` and `name[1]`. It is not a must here.

westonpace replied:
Ah, I see your point.  I was just porting over an existing test and I think both are probably ok so I think I'd prefer to leave it as is.

vibhatha replied:
Yes. It’s probably fine for this.

vibhatha commented on cpp/src/arrow/engine/substrait/extension_set.cc:
```diff
@@ -261,155 +389,390 @@ struct ExtensionIdRegistryImpl : ExtensionIdRegistry {
+  Status CanAddArrowToSubstraitCall(const std::string& function_name) const override {
+    if (arrow_to_substrait_.find(function_name) != arrow_to_substrait_.end()) {
+      return Status::Invalid(
+          "Cannot register function converter because a converter already exists");
```
should we add converter for `function_name` (add function_name in the message) already exists?

westonpace replied:
I updated this error message to:

```
      return Status::Invalid("Cannot register function converter for Substrait id ",
                             substrait_function_id.ToString(),
                             " because a converter already exists");
```

does that address your concern?

vibhatha replied:
Yes this works.

westonpace commented:
> General comment on usage of auto with ARROW_ASSIGN_OR_RAISE. There is a mix of auto and variable type through out the PR. Is there any specific reason? I think there was a recent discussion on minimizing usage of auto in certain places where it is hard to deterimine the data type for the readability sake. Is it the case here?

The primary places I use `auto` are `make_unique` and for STL iterators.  Looking through this PR I see there are also a number of places where I kept `auto` because I was just moving or refactoring existing code that was already using `auto`.

That being said, there may be a few places where I used `auto` incorrectly.  I'll take another scan.

westonpace commented:
@vibhatha / @rtpsw any other concerns?  I'm going to try and merge this tomorrow assuming it passes CI.

vibhatha commented on cpp/src/arrow/engine/substrait/expression_internal.cc:
```diff
@@ -41,6 +41,84 @@ namespace internal {
 using ::arrow::internal::make_unique;
 }  // namespace internal
 
+Status DecodeArg(const substrait::FunctionArgument& arg, int idx, SubstraitCall* call,
```
does `int idx` needs to be `uint32_t idx`?

westonpace replied:
I think I was originally aiming for compatibility with the source, which is a Substrait (i.e. protobuf) index which is signed.  But I agree it looks nicer with `uint32_t` since that matches the destination (and it seems I was inconsistent with casting).  I have updated it.

vibhatha commented on cpp/src/arrow/engine/substrait/function_test.cc:
```diff
@@ -0,0 +1,490 @@
+  return value_datum.make_array();
+}
+
+std::shared_ptr<Table> GetInputTable(
```
should this be instead `Result<std::shared_ptr<Table>>`?

vibhatha commented on cpp/src/arrow/engine/substrait/function_test.cc:
```diff
@@ -0,0 +1,490 @@
+  return table;
+}
+
+std::shared_ptr<Table> GetOutputTable(const std::string& output_value,
```
same here, should we use Result?

vibhatha commented on cpp/src/arrow/engine/substrait/function_test.cc:
```diff
@@ -0,0 +1,490 @@
+  return table;
+}
+
+std::shared_ptr<compute::ExecPlan> PlanFromTestCase(
```
Should we use `Result` here?

vibhatha commented on cpp/src/arrow/engine/substrait/function_test.cc:
```diff
@@ -0,0 +1,490 @@
+  return table;
+}
+
+std::shared_ptr<Table> GetOutputTableForAggregateCase(
```
Use `Result`?

vibhatha commented on cpp/src/arrow/engine/substrait/function_test.cc:
```diff
@@ -0,0 +1,490 @@
+  std::shared_ptr<DataType> output_type;
+};
+
+std::shared_ptr<Table> GetInputTableForAggregateCase(const AggregateTestCase& test_case) {
```
use `Result`?

vibhatha commented on cpp/src/arrow/engine/substrait/function_test.cc:
```diff
@@ -0,0 +1,490 @@
+  return table;
+}
+
+std::shared_ptr<compute::ExecPlan> PlanFromAggregateCase(
```
use `Result`?

vibhatha commented on cpp/src/arrow/engine/substrait/relation_internal.cc:
```diff
@@ -349,17 +365,20 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
-        } else {
-          return Status::Invalid(
-              "The grouping expression for an aggregate must be a direct reference.");
+      if (aggregate.groupings_size() > 0) {
```
Thanks for catching this.

vibhatha commented on cpp/src/arrow/engine/substrait/relation_internal.cc:
```diff
@@ -373,25 +392,14 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
             return Status::NotImplemented("Aggregate filters are not supported.");
           }
           const auto& agg_func = agg_measure.measure();
-          if (agg_func.arguments_size() != 1) {
-            return Status::NotImplemented("Aggregate function must be a unary function.");
-          }
-          int func_reference = agg_func.function_reference();
-          ARROW_ASSIGN_OR_RAISE(auto func_record, ext_set.DecodeFunction(func_reference));
-          // aggreagte function name
-          auto func_name = std::string(func_record.id.name);
-          // aggregate target
-          auto subs_func_args = agg_func.arguments(0);
-          ARROW_ASSIGN_OR_RAISE(auto field_expr, FromProto(subs_func_args.value(),
-                                                           ext_set, conversion_options));
-          auto target = field_expr.field_ref();
-          if (!target) {
-            return Status::Invalid(
-                "The input expression to an aggregate function must be a direct "
-                "reference.");
-          }
-          aggregates.emplace_back(compute::Aggregate{std::move(func_name), NULLPTR,
-                                                     std::move(*target), std::move("")});
```
Neat!

vibhatha commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -732,10 +737,9 @@ TEST(Substrait, ExtensionSetFromPlan) {
-    EXPECT_OK_AND_ASSIGN(auto decoded_add_func, ext_set.DecodeFunction(42));
-    EXPECT_EQ(decoded_add_func.id.uri, kArrowExtTypesUri);
-    EXPECT_EQ(decoded_add_func.id.name, "add");
-    EXPECT_EQ(decoded_add_func.name, "add");
```
did you mean to drop this one?

vibhatha replied:
`EXPECT_EQ(decoded_add_func.name, "add");`

westonpace replied:
`DecodeFunction` used to return an `Id` and an Arrow function name.  Returning only an Arrow function name was not sufficient.  Now it only returns an `Id`.  The job was split into `DecodeFunction` (which the extension set provides itself) and `GetSubstraitCallToArrow` (which the extension id registry provides).  So there is no `decoded_add_func.name` any longer.

vibhatha replied:
Got it. I missed that.

vibhatha commented on cpp/src/arrow/engine/substrait/test_plan_builder.h:
```diff
@@ -0,0 +1,54 @@
+namespace engine {
+namespace internal {
+
+Result<std::shared_ptr<Buffer>> CreateScanProjectSubstrait(
+    Id function_id, const std::shared_ptr<Table>& input_table,
+    const std::vector<std::string>& arguments,
+    const std::vector<std::shared_ptr<DataType>>& data_types,
+    const DataType& output_type);
+
+Result<std::shared_ptr<Buffer>> CreateScanAggSubstrait(
+    Id function_id, const std::shared_ptr<Table>& input_table,
+    const std::vector<int>& key_idxs, int arg_idx, const DataType& output_type);
```
should we add `docstring` here?

vibhatha commented on python/pyarrow/tests/test_substrait.py:
```diff
@@ -145,3 +145,8 @@ def test_binary_conversion_with_json_options(tmpdir):
+
+
+def test_get_supported_functions():
+    supported_functions = pa._substrait.get_supported_functions()
+    assert len(supported_functions) > 0
```
Is it wise to assert the supported functions by name rather than checking if it returns a `len > 0` list?

vibhatha reviewed (approved):
@westonpace Looks good to me. I added a few comments for minor changes. 
And thanks working on this and cleaning up the aggregate section with new test cases.

vibhatha commented:
@westonpace there is a link-error

```bash
unity_0_cxx.cxx.obj : error LNK2019: unresolved external symbol "class arrow::Result<class std::shared_ptr<class arrow::compute::ExecPlan> > __cdecl arrow::engine::DeserializePlan(class arrow::Buffer const &,class std::shared_ptr<class arrow::compute::SinkNodeConsumer> const &,class arrow::engine::ExtensionIdRegistry const *,class arrow::engine::ExtensionSet *,struct arrow::engine::ConversionOptions const &)" (?DeserializePlan@engine@arrow@@YA?AV?$Result@V?$shared_ptr@VExecPlan@compute@arrow@@@std@@@2@AEB
```
